tools/cli/adminElasticSearchCommands.go

```diff
@@ -231,3 +256,233 @@ func generateESDoc(msg *indexer.Message) map[string]interface{} {
+	var headers []string
+	var groupby, bucket map[string]interface{}
+	var buckets []interface{}
+	err = json.Unmarshal(*resp.Aggregations["groupby"], &groupby)
+	if err != nil {
+		ErrorAndExit("Fail to parse groupby", err)
+	}
+	buckets = groupby["buckets"].([]interface{})
+	if len(buckets) == 0 {
+		fmt.Println("no matching bucket")
+		return
+	}
+
+	// get the FIRST bucket in bucket list to extract all tags. These extracted tags are to be used as table heads
+	bucket = buckets[0].(map[string]interface{})
+	// record the column position in the table of each returned item
+	ids := make(map[string]int)
+	// We want these 3 columns shows at leftmost of the table in cadence report usage. It can be changed in future.
+	primaryCols := []string{"group_DomainID", "group_WorkflowType", "group_CloseStatus"}
+	primaryColsMap := map[string]int{
+		"group_DomainID":     1,
+		"group_WorkflowType": 1,
+		"group_CloseStatus":  1,
+	}
+	buckKeys := 0 // number of bucket keys, used for table collapsing in html report
+	if v, exist := bucket["key"]; exist {
+		vmap := v.(map[string]interface{})
+		// first search whether primaryCols keys exist, if found, put them at the table beginning
+		for _, k := range primaryCols {
+			if _, exist := vmap[k]; exist {
+				k = bucketKey(k) // trim the unnecessary prefix
+				headers = append(headers, k)
+				ids[k] = len(ids)
+				buckKeys++
+			}
+		}
+		// extract all remaining bucket keys
+		for k := range vmap {
+			if _, exist := primaryColsMap[k]; !exist {
+				k = bucketKey(k)
+				headers = append(headers, k)
+				ids[k] = len(ids)
+				buckKeys++
+			}
+		}
+	}
+	// extract all other non-key items and set the table head accordingly
+	for k := range bucket {
+		if k != "key" {
+			if k == "doc_count" {
+				k = "count"
+			}
+			headers = append(headers, k)
+			ids[k] = len(ids)
+		}
+	}
+	table.SetHeader(headers)
+
+	// read each bucket and fill the table, use map ids to find the correct spot
+	var tableData [][]string
+	for _, b := range buckets {
+		bucket = b.(map[string]interface{})
+		data := make([]string, len(headers))
+		for k, v := range bucket {
+			switch k {
+			case "key": // fill group key
+				vmap := v.(map[string]interface{})
+				for kk, vv := range vmap {
+					kk = bucketKey(kk)
+					data[ids[kk]] = fmt.Sprintf("%v", vv)
+				}
+			case "doc_count": // fill bucket size count
+				data[ids["count"]] = fmt.Sprintf("%v", v)
+			default:
+				var datum string
+				vm := v.(map[string]interface{})
+				if strings.Contains(k, "Attr_CustomDatetimeField") {
+					datum = fmt.Sprintf("%v", vm["value_as_string"])
+				} else {
+					datum = fmt.Sprintf("%v", vm["value"])
+					if strings.Contains(k, "Time") && !strings.Contains(k, "Attr_") {
+						datum, _ = timeStr(datum)
+					}
+				}
+				data[ids[k]] = datum
+			}
+		}
+		table.Append(data)
+		tableData = append(tableData, data)
+	}
+	table.Render()
+
+	// write html report
+	f, err := os.Create("report.html")
```

Comment:
try move writing to html and csv to separate functions, which will make code cleaner.